src/methods/widen.js

```diff
@@ -0,0 +1,309 @@
+    m1: PropertyBindings,
+    m2: PropertyBindings,
+    c1: CreatedObjects,
+    c2: CreatedObjects
+  ): PropertyBindings {
+    let widen = (b: PropertyBinding, d1: void | Descriptor, d2: void | Descriptor) => {
+      invariant(d1 !== undefined || d2 !== undefined, "widenMaps ensures that this cannot happen");
+      // If the PropertyBinding object has been freshly allocated do not widen (that happens in AbstractObjectValue)
+      if (d1 === undefined) {
+        if (b.object instanceof ObjectValue && c2.has(b.object)) return d2; // no widen
+        if (b.descriptor !== undefined && m1.has(b)) {
+          // property was present in (n-1)th iteration and deleted in nth iteration
+          d1 = cloneDescriptor(b.descriptor);
+          invariant(d1 !== undefined);
+          d1.value = realm.intrinsics.empty;
+        } else {
+          // no write to property in nth iteration, use the value from the (n-1)th iteration
+          d1 = b.descriptor;
+        }
+      }
+      if (d2 === undefined) {
+        if (b.object instanceof ObjectValue && c1.has(b.object)) return d1; // no widen
+        if (m2.has(b)) {
+          // property was present in nth iteration and deleted in (n+1)th iteration
+          d2 = cloneDescriptor(d1);
+          invariant(d2 !== undefined);
+          d2.value = realm.intrinsics.empty;
+        } else {
+          // no write to property in (n+1)th iteration, use the value from the nth iteration
+          d2 = d1;
+        }
+        invariant(d2 !== undefined);
+      }
+      return this.widenDescriptors(realm, d1, d2);
+    };
+    return this.widenMaps(m1, m2, widen);
+  }
+
+  widenDescriptors(realm: Realm, d1: void | Descriptor, d2: Descriptor): void | Descriptor {
+    if (d1 === undefined) {
+      // d2 is a property written to only in the (n+1)th iteration
+      return d2; // no widening needed. Note that another fixed point iteration will occur.
+    } else {
+      if (equalDescriptors(d1, d2)) {
+        if (!IsDataDescriptor(realm, d1)) return d1; // identical accessor properties need not be widened.
+        let dc = cloneDescriptor(d1);
+        invariant(dc !== undefined);
+        dc.value = this.widenValues(realm, d1.value, d2.value);
+        return dc;
+      }
+      //todo: #1174 if we get here, the loop body contains a call to create a property and different iterations
+      // create them differently. That seems beyond what a fixpoint computation can reasonably handle without
+      // losing precision. Report an error here.
+      throw new FatalError();
+    }
+  }
+
+  // If e2 is the result of a loop iteration starting with effects e1 and it has the same elements as e1,
+  // then we have reached a fixed point and no further calls to widen are needed. e1/e2 represent a general
+  // summary of the loop, regardless of how many iterations will be performed at runtime.
+  equalsEffects(e1: Effects, e2: Effects): boolean {
+    let [result1, , bindings1, properties1] = e1;
```

Comment:
When pattern matching I like to name all constituents, even if unused. Makes it clear that its not a typo.

Reply:
I am in two minds about this. Giving it a name help readability, but then you need to add an artificial reference to it in order to shut up the "variable not used" lint warning. The latter clutters up the code and decreases readability. In OCaml, people often just use "_" in this situation (although "_name" is probably the better option). I actually don't think it is reasonable to look at the pattern and suspect that it is a typo. In general, the type checker will be quick to notice such typos, so human readers can afford to be a bit more relaxed.

Reply:
Fair enough and while I agree that a SML type checker would catch such a typo, I am less convinced that JavaScript will.

Reply:
Flow is generally able to catch such errors.